### ISSUE TITLE
Linux jails

### ISSUE THREAD

**Issue by Juronski** [cbsd/cbsd#589]
Mandatory info for bug reports:
FreeBSD version ( uname -a ):  FreeBSD 12.2-RELEASE-p1 GENERIC  amd64
CBSD version ( cbsd version ):  13.0.4

Hi CBSD, 

Thank you for the Linux Jails Profiles. Have a question about it please. 

Not all changes made in a Linux jail are saved on reboot/restart jail. Is that expected behaviour? 

J.

**Comment by olevole** [comment]
Unfortunately, I can't comment without any details. CBSD profile uses information/scenarios from https://wiki.freebsd.org/linuxjails pages.  I also see that this page does not write anything about the autorun service. Usually Linux services are launched by SystemD. However, it does not work in Linux Jail. Perhaps other problems are also caused with the systemd. I do not know what the FreeBSD can offer here. And I do not know what we can do in CBSD

**Comment by Juronski** [comment]
I can give some examples. I have created a Buster Linux Jail with default (CBSD) settings. When packages are installed apt seems to forget about it after restart of the jail. After restart the installed packages are there but apt has no record of it. And after restart of Buster Jail changes made in (apt) sources.list are not saved. Also in the CentOS7 jail not all changes are saved. For example users. But Yum does not forget about installed packages after restart of the CentOS jail. When I create a linux jail from a chroot without CBSD this does not happen. On CBSD the changes need to be made (with chroot) in the skeleton to make them permanent in the Buster Jail. On the one hand changes are saved after restart like the installed packages, but on the other hand changes in configuration-files and e.g. apt-database are not saved. Are these examples enough details? Unfortunately there is no log of this. 

The things you mention with auto run service are expected in a linux jail. Buster Jail is not booted with systemd. CentOS Jail uses systemd.

**Comment by Juronski** [comment]
I think this has to do with the fact that when a Linux Jail is started in CBSD, CBSD applies custom skel dir template. So when a Buster Jail is started the jail populates from: /usr/jails/share/FreeBSD-jail-debian-buster-system-skel. 

Is it possible to turn this off after creating the Linux Jail to see if that solves the issue?

**Comment by olevole** [comment]
Yes, it seems that erroneous behavior. I'll see later why this happens every time. It should work only once

**Comment by Juronski** [comment]
Ok. Thank you.

**Comment by olevole** [comment]
@Juronski Yes, for some reasons I did not check that the content already exists. 
Changes ( in all **tests/00.check_distribution.sh** ) must correct wrong behavior: 
```
-- . ${distdir}/freebsd_world.subr
-- customskel
++ if [ ! -r ${data}/bin/bash ]; then
++	. ${distdir}/freebsd_world.subr
++	customskel
++ fi
```
But you need to remove old cache dir first: `rm -rf ~cbsd/share/FreeBSD-jail-*-system-skel` and re-run `cbsd initenv`
Let me know if the problem still remains after that.

**Comment by Juronski** [comment]
Hi @Olevole,

The problem still remains on my side. This is what I did:		

1. Copied the new FreeBSD-jail-*-system-skel from github to /usr/local/cbsd/share/FreeBSD-jail- * -system-skel
2. Removed the old /usr/jail/share/FreeBSD-jail-*-system-skel `rm -rf /usr/jail/share/FreeBSD-jail-*-system-ske`l
3. Did: `sudo cbsd initenv`
4. Started the jail (in this case Buster Jail)

Then I get:

> Shall i download distribution via deboostrap from http://deb.debian.org/debian/?
> [yes(1) or no(0)]

Answer Yes (1)

> debootstrap --include=openssh-server,locales,rsync,sharutils,psmisc,patch,less,apt --components main,contrib buster /usr/jails/share/FreeBSD-jail-debian-buster-system-skel http://deb.debian.org/debian/
> etc

 /usr/jail/share/FreeBSD-jail-buster-system-skel is created

And with the next stop/start I get again:

> Check environment script: 00.check_distribution.sh
> Applying custom skel dir template from: /usr/jails/share/FreeBSD-jail-debian-buster-system-skel

This is the same on CentOS7 Jail. Also created a new Linux Jail to test.

**Comment by olevole** [comment]
It is not very clear why the check does not work for you. Look at the contents of the file:
```
cat ~cbsd/share/FreeBSD-jail-debian-buster-system-skel/tests/00.check_distribution.sh 
```

it should match this: https://github.com/cbsd/cbsd/blob/develop/share/FreeBSD-jail-debian-buster-system-skel/tests/00.check_distribution.sh

The rewriting happens in the **customkel** call, which is executed if there is no bash inside the container:
 ```
 if [ ! -r ${data}/bin/bash ]; then
	. ${distdir}/freebsd_world.subr
	customskel
fi
```

**Comment by Juronski** [comment]
Have redone everything according the instructions above, rebooted and now it works for the Debian Buster Jail. Do not know why it did not work before.

Thank you so much @olevole. Verry nice to have Linux Jails on CBSD.

**Comment by olevole** [comment]
I think I'll change the logic in the near future - the contrib scripts (/usr/local/cbsd/share/*) will be used by default. 
And in the working directory (~cbsd/share/* ) there will be a place for custom scripts if users want to change them (custom).

**Comment by Juronski** [comment]
Yes. All kind of scripts can be made for customized Linux Jails. And a way need to be found to get vnet to work from within Linux Jails. Think that can help solve general Linux Jail issues like: 
> socket() [::]:80 failed (93: Protocol not supported)


`

**Comment by olevole** [comment]
![deb_vnet](https://user-images.githubusercontent.com/926409/116308946-d1bf7400-a7b0-11eb-9c08-47fd2967c73e.png)

@Juronski do you mean that you cannot set the IP address inside the **vnet** container? 
i started refactoring linux container support and the latest github CBSD version ready for debian buster. Also vnet-based jails created successful

**Comment by olevole** [comment]
As far as I understand, there is a issue with the Linux netlink subsystem, but via sysfs i can see my NIC:

```
root@deb:~# ls -la /sys/class/net/
total 0K
dr-xr-xr-x 1 root root 0 Apr 27 20:25 .
dr-xr-xr-x 1 root root 0 Apr 27 20:25 ..
dr-xr-xr-x 1 root root 0 Apr 27 20:25 eth0
dr-xr-xr-x 1 root root 0 Apr 27 20:25 lo


root@deb:~# cat /sys/class/net/eth0/address 
00:a0:98:de:e1:70
```

**Comment by olevole** [comment]
@Juronski I don't know anything about Linuxulator + netlink support, but it might be possible to set the address without it, for example: https://stackoverflow.com/questions/6652384/how-to-set-the-ip-address-from-c-in-linux/6652532  ;-)

I think you should go to **freebsd-emulation** maillist now.

**Comment by Juronski** [comment]
@olevole yes, I can not get an IP address inside the container/jail. I get eth0 (same as you), but inside the jail I do not get the eth0 to work. Get the same result when I do `ls -la /sys/class/net/`. I go the freebsd-emulation maillist now.  

Thanks again.

**Comment by olevole** [comment]
BTW, you can use native FreeBSD ioctl (eg ifconfig/route) to configure the Linux container network in vnet jail. Nevertheless, I do not understand what the virtual stack gives you in linux jail, because all utilities are highly dependent on the hard components of the Linux kernel (netlink, systemd). I recorded a funny video of how it might look (sorry, there is Russian speech): https://youtu.be/wDox97l3fcQ :-)

**Comment by gizahNL** [comment]
For my Docker port I create a freebsd vnet jail with root fs set to /, configure the network in that and then create a child jail in that jail for linux and set it up to inherit the networking.

**Comment by olevole** [comment]
> For my Docker port I create a freebsd vnet jail with root fs set to /, configure the network in that and then create a child jail in that jail for linux and set it up to inherit the networking.

This is also a difficult scheme)

I just added into 'bootstrap script'  copying of '/sbin/ifconfig' and '/sbin/route' from FreeBSD host to Linux skel directory + all libs ( ldd \`which ifconfig route\` ).

**Comment by gizahNL** [comment]
> > For my Docker port I create a freebsd vnet jail with root fs set to /, configure the network in that and then create a child jail in that jail for linux and set it up to inherit the networking.
> 
> This is also a difficult scheme)
> 
> I just added into 'bootstrap script' copying of '/sbin/ifconfig' and '/sbin/route' from FreeBSD host to Linux skel directory + all libs ( ldd `which ifconfig route` ).

Yeah that also works ;) I played around as well with static compilation of ifconfig & route tools, that also worked. But it copying in files does not fit with the design philosophy of docker (but it is indeed perfectly fine for jails orchestrated via CBSD)
